File: Mathematics/modular exp/cpp/modular_exp.cpp
#include <bits/stdc++.h>
using namespace std;

/// modn is the chosen prime
const long long modn = 1000000007;
inline long long mod(long long x) { return x % modn; }

long long expo(long long num, long long exp){
	long long ans = 1;
	while(exp){
		if(exp&1)
			ans = mod(ans*num);
		num = mod(num*num);
		exp >>= 1;
	}
	return ans;
}

// computes x's modular inverse
long long mod_inv(long long x){
	return expo(x, modn-2);
}

int main (){
	long long x, exp;
	scanf("%lld %lld", &x, &exp);
	printf("%lld\n", expo(x, exp));
}

File: Mathematics/modular exp/java/ModularExp.java
public class ModularExp {
	// computes n raised to the nonnegative power e modulo m
	public static long exp(long n, int e, int m) {
		if (e == 0) {
			return 1L;
		}

		n %= m;
		long ans = 1;

		if (e%2 == 0) {
			return exp(ans*ans, e/2, m);
		}
		else {
			return n*exp(ans*ans, e/2, m);
		}
	}
	
	public static void main(String[] args) {
		long n = 137;
		int e = 207;
		int m = 18;
		System.out.println(exp(n,e,m));
	}
}
